Swap GA graduation: blog-post (#48099)
kubernetes/website: content/en/blog/_posts/2024-09-25-swap-ga-graduation.md

> 

Reviewer: ```suggestion
```

---
layout: blog
title: "Kubernetes 1.32: GA support for using swap on Linux"
date: 2024-09-25T10:00:00-08:00
slug: swap-linux-ga

> ---
> 
> **Author:** Itamar Holder (Red Hat)

Reviewer: ```suggestion
author: >
  Itamar Holder (Red Hat)
---
```

Swap is a fundamental and an invaluable Linux feature.
It offers numerous benefits, such as effectively increasing a node’s memory by
swapping out unused data,
shielding nodes from system-level memory spikes,
preventing Pods from crashing when they hit their memory limits,
and [much more](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#user-stories).
As a result, the node special interest group within the Kubernetes project
has invested significant effort into supporting swap on Linux nodes as a GA feature.

> The 1.22 release [introduced Alpha support](/blog/2021/08/09/run-nodes-with-swap-alpha/)

Reviewer: ```suggestion
The 1.22 release [introduced](/blog/2021/08/09/run-nodes-with-swap-alpha/) Alpha support
```

Now, in the 1.32 release, swap is graduating to be Generally Available.

Prior to version 1.22, Kubernetes did not provide support for swap memory on Linux systems.
This was due to the inherent difficulty in guaranteeing and accounting for pod memory utilization
when swap memory was involved. As a result, swap support was deemed out of scope in the initial
design of Kubernetes, and the default behavior of a kubelet was to fail to start if swap memory
was detected on a node.

In version 1.22, the swap feature for Linux was initially introduced in its Alpha stage.
This provided Linux users the opportunity to experiment with the swap feature for the first time.
However, as an Alpha version, it was not fully developed and only partially worked on limited environments.

In version 1.28 swap support on Linux nodes was promoted to Beta.
The Beta version was a drastic leap forward.
Not only did it fix a large amount of bugs and made swap work in a stable way,
but it also brought cgroup v2 support, introduced a wide variety of tests
which include complex scenarios such as node-level pressure, and more.
It also brought many exciting new capabilities such as the `LimitedSwap` behavior

> which sets an auto-calculated swap limit to containers, metrics support for tools
> like Prometheus (through the `/metrics/resource` endpoint) and Summary API for

Reviewer: Rather than list a specific (albeit popular) implementation, are we able to write:
```suggestion
which sets an auto-calculated swap limit to containers, OpenMetrics intrumentation
support (through the `/metrics/resource` endpoint) and Summary API for
```
?

> Vertical Pod Autoscalers (through the `/stats/summary` endpoint), and more.

Reviewer: ```suggestion
VerticalPodAutoscalers (through the `/stats/summary` endpoint), and more.
```

## How do I use it?

In order for the kubelet to initizalize on a swap-enabled node, the `failSwapOn`
field must be set to `false` on kubelet's configuration setting, or the deprecated
`--fail-swap-on` command line flag must be deactivated.

It is possible to configure the `memorySwap.swapBehavior` option to define the
manner in which a node utilizes swap memory.
For instance,

```yaml
# this fragment goes into the kubelet's configuration file
memorySwap:
  swapBehavior: LimitedSwap
```

The currently available configuration options for `swapBehavior` are:
- `NoSwap` (default): Kubernetes workloads cannot use swap. However, processes
  outside of Kubernetes' scope, like system daemons (such as kubelet itself!) can utilize swap.
  This behavior is beneficial for protecting the node from system-level memory spikes,
  but it does not safeguard the workloads themselves from such spikes.
- `LimitedSwap`: Kubernetes workloads can utilize swap memory, but with certain limitations.
  The amount of swap available to a Pod is determined automatically,
  based on the proportion of the memory requested relative to the node's total memory.

>   Only non-high-priority Pods under the [Burstable](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#burstable)

Reviewer: ```suggestion
  Only non-high-priority Pods under the [Burstable](/docs/concepts/workloads/pods/pod-qos/#burstable)
```

  Quality of Service (QoS) tier are permitted to use swap.
  For more details, see the [section below](#how-is-the-swap-limit-being-determined-with-limitedswap).

If configuration for `memorySwap` is not specified,
by default the kubelet will apply the same behaviour as the `NoSwap` setting.

> Note that `NodeSwap` is supported for **cgroup v2** only.

Reviewer: ```suggestion
On Linux nodes, Kubernetes only supports running with swap enabled for hosts that use cgroup v2.
```

`(containerMemoryRequest / nodeTotalMemory) × totalPodsSwapAvailable`

In other words, the amount of swap that a container is able to use is proportionate to its
memory request, the node's total physical memory and the total amount of swap memory on
the node that is available for use by Pods.

It is important to note that, for containers within Burstable QoS Pods, it is possible to
opt-out of swap usage by specifying memory requests that are equal to memory limits.
Containers configured in this manner will not have access to swap memory.

## How does it work?

There are a number of possible ways that one could envision swap use on a node.
When swap is already provisioned and available on a node,
the kubelet is able to be configured so that:
- It can start with swap on.
- It will direct the Container Runtime Interface to allocate zero swap memory
  to Kubernetes workloads by default.

Swap configuration on a node is exposed to a cluster admin via the
[`memorySwap` in the KubeletConfiguration](/docs/reference/config-api/kubelet-config.v1).
As a cluster administrator, you can specify the node's behaviour in the
presence of swap memory by setting `memorySwap.swapBehavior`.

> The kubelet [employs the CRI](https://kubernetes.io/docs/concepts/architecture/cri/)

Reviewer: ```suggestion
The kubelet employs the [CRI](https://kubernetes.io/docs/concepts/architecture/cri/)
```

back to memory is a heavy operation, sometimes slower by many orders of magnitude,
which can cause unexpected performance regressions.
Furthermore, swap changes a system's behaviour under memory pressure.
Since enabling swap permits greater memory usage for workloads in Kubernetes that
cannot be predictably accounted for, it also increases the risk of noisy neighbours
and unexpected packing configurations, as the scheduler currently does not account for swap memory usage.

The performance of a node with swap memory enabled depends on the underlying physical storage.
When swap memory is in use, performance will be significantly worse in an I/O
operations per second (IOPS) constrained environment, such as a cloud VM with
I/O throttling, when compared to faster storage mediums like solid-state drives
or NVMe.
As swap might cause IO pressure, it is recommended to give a higher IO latency
priority to system critical daemons. See the relevant section in the
[best-practices](#best-practices) section below.

### Memory-backed volumes

On Linux nodes, memory-backed volumes (like [Secrets](/docs/concepts/configuration/secret/)

> and [emptyDirs](/docs/concepts/storage/volumes/#emptydir) with `medium: Memory`)

Reviewer: ```suggestion
On Linux nodes, memory-backed volumes (such as [`secret`](/docs/concepts/configuration/secret/)
volume mounts, or or [`emptyDir`](/docs/concepts/storage/volumes/#emptydir) with `medium: Memory`)
```

> are implemented with as `tmpfs` mount.

Reviewer: ```suggestion
are implemented with a `tmpfs` filesystem.
```

The contents of such volumes should remain in memory at all times, hence should
not be swapped to disk.
To ensure the contents of such volumes remain in memory, the `noswap` tmpfs option
is being used.

The Linux kernel officially supports the `noswap` option from version 6.3 (more info

> can be found [here](/docs/reference/node/kernel-version-requirements/#requirements-other)).

Reviewer: ```suggestion
can be found in [Linux Kernel Version Requirements](/docs/reference/node/kernel-version-requirements/#requirements-other)).
```

> However, the different distributions often choose to backport this option to older

Reviewer: ```suggestion
However, the different distributions often choose to backport this mount option to older
```

Linux versions as well.

> In order to understand if the node supports the `noswap` option, Kubelet will do the following:

Reviewer: How about:
```
In order to verify whether the node supports the `noswap` option, the kubelet will do the following:
```

* If the kernel's version is above 6.3 then the `noswap` option will be assumed to be supported.
* Otherwise, kubelet would try to mount a dummy tmpfs with the `noswap` option at startup.
  If kubelet fails with an error indicating of an unknown option, `noswap` will be assumed

>   to not be supported, hence will not be used.

Reviewer: What happens once the kubelet detects the lack of `noswap` support? Does the kubelet refuse to start? Do you get an Event against the node?

Reviewer: I'd kind of hope to see a kubelet configuration option such as `allowInsecureMemoryVolumes` that you have to turn on if using an older kernel. Is there anything like that?

Author: > What happens once the kubelet detects the lack of `noswap` support? Does the kubelet refuse to start? Do you get an Event against the node?

When `noswap` is not supported, kubelet would start with a log entry warning.

There were long discussions regarding this topic (for example, look [here](https://github.com/kubernetes/kubernetes/issues/105978#issuecomment-2078246497) for more information). In short, this behavior is reasonable since most distribution backported this mount option to much older versions. The vast majority of systems would have this supported, and there is documentation and warnings for the systems that don't, that anyhow are likely to update their Kernel version in a reasonable timeframe.

Reviewer: We might want to call out the risk of enabling swap on an old kernel; even if it's unlikely, some people may face important opportunities to get things really wrong. Helping them they that they haven't and emphasizing how important it is to check? I recommend it.

Reviewer: > When `noswap` is not supported, kubelet would start with a log entry warning.

OK, but: don't tell the reviewers. Tell the readers.

Author: > Helping them they that they haven't and emphasizing how important it is to check?

So you suggest adding a node-level event in addition to the kubelet log? I think that might be reasonable.

Author: > > When `noswap` is not supported, kubelet would start with a log entry warning.
> 
> OK, but: don't tell the reviewers. Tell the readers.

Added.

Reviewer: The Node-level event would be a nice improvement, but whether it's in or not without it let's clearly call out how to check what's going on.

  If kubelet succeeds, the dummy tmpfs will be deleted and the `noswap` option will be used.
  * If the `noswap` option is not supported, kubelet will emit a warning log entry,
    then continue its execution.

It is deeply encouraged to encrypt the swap space.
See the [section above](#setting-up-encrypted-swap) with an example for setting unencrypted swap.
However, handling encrypted swap is not within the scope of kubelet;
rather, it is a general OS configuration concern and should be addressed at that level.
It is the administrator's responsibility to provision encrypted swap to mitigate this risk.

> ## Best practices

Reviewer: ```suggestion
## Good practice for using swap in a Kubernetes cluster
```

### Disable swap for system-critical daemons

During the testing phase and based on user feedback, it was observed that the performance
of system-critical daemons and services might degrade.
This implies that system daemons, including the kubelet, could operate slower than usual.
If this issue is encountered, it is advisable to configure the cgroup of the system slice
to prevent swapping (i.e., set `memory.swap.max=0`).

> ### Protect system-critical daemons for io-latency

Reviewer: ```suggestion
### Protect system-critical daemons for I/O latency
```

Swap can increase the I/O load on a node.
When memory pressure causes the kernel to rapidly swap pages in and out,
system-critical daemons and services that rely on I/O operations may
experience performance degradation.

> To mitigate this, it is recommended to prioritize the system slice in terms of I/O latency.
> This can be achieved by setting `io.latency` for the system slice,
> thereby granting it higher I/O priority.

Reviewer: ```suggestion
To mitigate this, it is recommended to prioritize the systemd slice named `system` in terms of I/O latency.
This can be achieved by setting `io.latency` for the system slice,
thereby granting it higher I/O priority.
```
Does this work, though? What is it being prioritized relative to?

Author: Yeah it does. As the docs say:
```
Specifies the number of milliseconds a process can wait before IO from other processes is given to it.
...
effectively prioritizing the job with the lowest io.latency setting.
```

See https://facebookmicrosites.github.io/cgroup2/docs/io-controller.html#protecting-workloads-with-iolatency for more info regarding this.

Author: Regarding the suggested change: I wanted to avoid mentioning systemd specifically because some environments might run without systemd, but still have a dedicated cgroup hierarchy for system processes.

WDYT?

Reviewer: What other slice do you think is being deemphasized here? I worry that making that change won't actually improve things on a typical node.

Author: > What other slice do you think is being deemphasized here? 

The kubepods slice.

> I worry that making that change won't actually improve things on a typical node.

We tested this pretty extensively. I believe that in certain situations it is extremely helpful. For example, in situations where pods were swapping in-and-out extensively causing an IO pressure for system processes that are outside of kubelet's scope.

However if you think it's better to remove this section I won't insist. This is just a recommendation after all :)

Reviewer: As far as I know, [in the Linux mainstream] only systemd has a slice concept. So either write for systemd users, or make it even more generic.

Author: Done.
Wrote:
```
To mitigate this, it is recommended for systemd users to prioritize the system slice in terms of I/O latency.
For non-systemd users,
setting up a dedicated cgroup for system daemons and processes and prioritizing I/O latency in the same way is advised.
This can be achieved by setting `io.latency` for the system slice,
thereby granting it higher I/O priority.
See [cgroup's documentation](https://www.kernel.org/doc/Documentation/admin-guide/cgroup-v2.rst) for more info.
```

> ### Control Plane Swap

Reviewer: ```suggestion
### Swap and control plane nodes
```

> It is recommended to enable swap only for the worker nodes.

Reviewer: ```suggestion
The Kubernetes project recommends running control plane nodes without any swap space configured.
```

The control plane primarily hosts Guaranteed QoS Pods, so swap can generally be disabled.
The main concern is that swapping critical services on the control plane could negatively impact performance.

### Use of a dedicated disk for swap

It is recommended to use a separate, encrypted disk for the swap partition.
If swap resides on a partition or the root filesystem, workloads may interfere
with system processes that need to write to disk.
When they share the same disk, processes can overwhelm swap,
disrupting the I/O of kubelet, container runtime, and systemd, which would impact other workloads.
Since swap space is located on a disk, it is crucial to ensure the disk is fast enough for the intended use cases.

## Looking ahead

> As mentioned above, the NodeSwap feature introduces basic swap support,

Reviewer: How about:
```suggestion
As mentioned above, this stable release provides basic but generally available support
for swap on Linux nodes. It's a foundational implementation
```

for configuring swap memory usage for Kubernetes workloads running on Linux on a per-node basis.
Later, in release 1.28, support for swap on Linux nodes has graduated to Beta, along with many
new improvements.
Now, in the 1.32 release, swap is graduating to be Generally Available.

Prior to version 1.22, Kubernetes did not provide support for swap memory on Linux systems.
This was due to the inherent difficulty in guaranteeing and accounting for pod memory utilization
when swap memory was involved. As a result, swap support was deemed out of scope in the initial
design of Kubernetes, and the default behavior of a kubelet was to fail to start if swap memory
was detected on a node.

In version 1.22, the swap feature for Linux was initially introduced in its Alpha stage.
This provided Linux users the opportunity to experiment with the swap feature for the first time.
However, as an Alpha version, it was not fully developed and only partially worked on limited environments.

In version 1.28 swap support on Linux nodes was promoted to Beta.
The Beta version was a drastic leap forward.
Not only did it fix a large amount of bugs and made swap work in a stable way,

> but it also brought cgroup v2 support, introduced a wide variety of tests

Reviewer 2: didn't we disable it for cgroupv1?

Author: Yes, we eventually did. Do you think I should mention that?
Initially (during alpha) only v1 was supported.

which include complex scenarios such as node-level pressure, and more.
It also brought many exciting new capabilities such as the `LimitedSwap` behavior

> which sets an auto-calculated swap limit to containers, OpenMetrics instrumentation

Reviewer 2: OpenMetrics is archived, we are back to call it Prometheuse I think: https://horovits.medium.com/openmetrics-is-archived-merged-into-prometheus-d555598d2d04

Containers configured in this manner will not have access to swap memory.

## How does it work?

There are a number of possible ways that one could envision swap use on a node.
When swap is already provisioned and available on a node,
the kubelet is able to be configured so that:
- It can start with swap on.
- It will direct the Container Runtime Interface to allocate zero swap memory
  to Kubernetes workloads by default.

Swap configuration on a node is exposed to a cluster admin via the
[`memorySwap` in the KubeletConfiguration](/docs/reference/config-api/kubelet-config.v1).
As a cluster administrator, you can specify the node's behaviour in the
presence of swap memory by setting `memorySwap.swapBehavior`.

The kubelet employs the [CRI](https://kubernetes.io/docs/concepts/architecture/cri/)
(container runtime interface) API, and directs the container runtime to
configure specific cgroup v2 parameters (such as `memory.swap.max`) in a manner that will
enable the desired swap configuration for a container. For runtimes that use control groups,
the container runtime is then responsible for writing these settings to the container-level cgroup.

## How can I monitor swap?

> ### Node-level metric statistics

Reviewer 2: do we have per-pod or per container stats?

Author: Yeah, good catch! These stats include container-level stats, I've done some re-wording to emphasize that.

BTW, I think that what confused me is that the docs use this terminology at https://kubernetes.io/docs/reference/instrumentation/node-metrics/

Kubelet now collects
[node-level metric statistics](/docs/reference/instrumentation/node-metrics/),
which can be accessed at the `/metrics/resource` (which is used mainly by monitoring
tools like Prometheus) and `/stats/summary` (which is used mainly by Autoscalers) kubelet HTTP endpoints.
This allows clients who can directly interrogate the kubelet to
monitor swap usage and remaining swap memory when using `LimitedSwap`.
Additionally, a `machine_swap_bytes` metric has been added to cadvisor to show
the total physical swap capacity of the machine.

> ### NFD (Node Feature Discovery)

Reviewer: ```suggestion
### Node Feature Discovery (NFD) {#node-feature-discovery}
```

## Caveats

Having swap available on a system reduces predictability.
While swap can enhance performance by making more RAM available, swapping data
back to memory is a heavy operation, sometimes slower by many orders of magnitude,
which can cause unexpected performance regressions.
Furthermore, swap changes a system's behaviour under memory pressure.
Since enabling swap permits greater memory usage for workloads in Kubernetes that
cannot be predictably accounted for, it also increases the risk of noisy neighbours
and unexpected packing configurations, as the scheduler currently does not account for swap memory usage.

The performance of a node with swap memory enabled depends on the underlying physical storage.
When swap memory is in use, performance will be significantly worse in an I/O
operations per second (IOPS) constrained environment, such as a cloud VM with
I/O throttling, when compared to faster storage mediums like solid-state drives
or NVMe.
As swap might cause IO pressure, it is recommended to give a higher IO latency
priority to system critical daemons. See the relevant section in the

> [recommended practices](#good-practice-for-using-swap-in-a-kubernetes-cluster) section below.

Reviewer 2: Also mention that there is a big risk of a noisy neighbour when other Pods which use their RAM more often may force other processes into swapping

However, the different distributions often choose to backport this mount option to older
Linux versions as well.

In order to verify whether the node supports the `noswap` option, the kubelet will do the following:
* If the kernel's version is above 6.3 then the `noswap` option will be assumed to be supported.
* Otherwise, kubelet would try to mount a dummy tmpfs with the `noswap` option at startup.
  If kubelet fails with an error indicating of an unknown option, `noswap` will be assumed
  to not be supported, hence will not be used.
  A kubelet log entry will be emitted to warn the user about memory-backed volumes might swap to disk.
  If kubelet succeeds, the dummy tmpfs will be deleted and the `noswap` option will be used.
  * If the `noswap` option is not supported, kubelet will emit a warning log entry,
    then continue its execution.

It is deeply encouraged to encrypt the swap space.
See the [section above](#setting-up-encrypted-swap) with an example for setting unencrypted swap.
However, handling encrypted swap is not within the scope of kubelet;
rather, it is a general OS configuration concern and should be addressed at that level.

> It is the administrator's responsibility to provision encrypted swap to mitigate this risk.

Reviewer 2: Can we add some numbers on performance impact of encrypted vs. non-encrypted

eviction manager identify memory spikes in time so it can execute custom logic and take care of reclaiming
memory and evicting pods before the kernel's OOM killer is invoked.

However, with swap enabled, the situation is fundamentally different.
Since swapping is a heavy operation in terms of performance, the kernel will try avoiding
it as long as it can, and will usually start swapping only when it lacks physical memory.

Therefore, when swap is enabled, it is acceptable to use all the available RAM which
will result in the node using some of the swap memory.

#### Current eviction limitations

For both Linux and Windows nodes, the eviction manager is not aware of how much of the swap storage is used.

Let X be the memory threshold below which the kernel initiates swapping.
This means that most probably either one of the two behaviors will take place:
1. Memory evictions threshold > X: In this scenario, kubelet's eviction manager would kick in before the kernel has
   a chance to swap memory.
   This means that the kernel would generally not be able to swap memory (unless the memory spike
   is fast enough for the eviction manager to handle).

> 2. Memory evictions threshold <= X: In this scenario, the kernel would start swapping memory before the memory threshold 

Reviewer 2: Is there a third point - when we already exhausted the whole swap and evictions will kick in?

Author: This section is not relevant anymore and is now removed.

   would be met. This practically means that the eviction manager would never kick in, hence it is practically turned off (unless
   for extreme cases where the memory spike is faster than the kernel's ability to swap memory). 

The cluster admin can choose which of the above is the desired behavior.
An advanced cluster admin can use the approach outlined [below](#advanced-best-practices-for-manually-setting-memory-evictions).

> #### Advanced good practice for manually setting memory evictions

Reviewer 2: do we want to recommend disabling eviction?

Later, when the _min_ threshold is reached, the kernel goes back to "indirect reclaim", then when the _high_ threshold
is met the node is not considered under pressure anymore.

{{< mermaid >}}
sequenceDiagram
    Below High->>+High: No memory pressure.
    High->>+Min: Node isn’t under pressure.
    Min->>+Low: Indirect reclaim: kswapd starts to asynchronously reclaim memory.
    Low->>+Above Low: Direct reclaim: kswapd aggressively reclaims memory. Memory is throttled.
    Above Low-->>+Low: Direct reclaim continues with memory throttle.
    Low-->>+Min: Back to indirect reclaim, with no memory throttle.
    Min-->>+High: Back to normal.
{{< /mermaid >}}

To figure out what's the value of the different watermarks, one can do the following (the output is trimmed
for simplicity. values are in memory pages):
```shell
> cat /proc/zoneinfo
--
Node 0, zone    DMA32
        min      569
        low      1120
        high     1671
--
Node 0, zone   Normal
        min      16322
        low      32138
        high     47954
--
...
```

This is a real example from a 64GB RAM machine with the default kernel configurations. On such a machine:

> * The `high` watermark is at `187.32Mi`.
> * The `low` watermark is at `125.54Mi`.
> * The `min` watermark is at `63.76Mi`.

Reviewer: I'd include the unit:
```suggestion
* The `high` watermark is at `187.32MiB`.
* The `low` watermark is at `125.54MiB`.
* The `min` watermark is at `63.76MiB`.
```

system-critical daemons and services that rely on I/O operations may
experience performance degradation.

To mitigate this, it is recommended to prioritize the system slice in terms of I/O latency.
This can be achieved by setting `io.latency` for the system slice,
thereby granting it higher I/O priority.

### Swap and control plane nodes

The Kubernetes project recommends running control plane nodes without any swap space configured.
The control plane primarily hosts Guaranteed QoS Pods, so swap can generally be disabled.
The main concern is that swapping critical services on the control plane could negatively impact performance.

### Use of a dedicated disk for swap

It is recommended to use a separate, encrypted disk for the swap partition.
If swap resides on a partition or the root filesystem, workloads may interfere
with system processes that need to write to disk.
When they share the same disk, processes can overwhelm swap,
disrupting the I/O of kubelet, container runtime, and systemd, which would impact other workloads.
Since swap space is located on a disk, it is crucial to ensure the disk is fast enough for the intended use cases.

> 

Reviewer: Alternative: configure I/O priorities between different mapped areas of one backing device.
